service/src/main/java/bio/terra/workspace/service/iam/SamService.java

```diff
@@ -670,46 +695,56 @@ private String syncPolicyOnObject(
    * Create a controlled resource in Sam.
    *
    * @param resource The WSM representation of the resource to create.
-   * @param privateIamRoles The IAM role(s) to grant a private user. Required for private resources,
-   *     should be null otherwise.
-   * @param assignedUserEmail Email identifier of the assigned user of this resource. Required for
-   *     private resources, should be null otherwise.
+   * @param privateIamRole The IAM role to grant on a private resource. It is required for
+   *     user-private resources and optional for application-private resources.
+   * @param assignedUserEmail Email identifier of the assigned user of this resource. Same
+   *     constraints as privateIamRoles.
    * @param userRequest Credentials to use for talking to Sam.
    */
   @Traced
   public void createControlledResource(
       ControlledResource resource,
-      @Nullable List<ControlledResourceIamRole> privateIamRoles,
+      @Nullable ControlledResourceIamRole privateIamRole,
       @Nullable String assignedUserEmail,
       AuthenticatedUserRequest userRequest)
       throws InterruptedException {
-    // Set up the master OWNER user in Sam for all controlled resources, if it's not already.
+    // We need the WSM SA for setting controlled resource policies
     initializeWsmServiceAccount();
-    ResourcesApi resourceApi = samResourcesApi(userRequest.getRequiredToken());
+
     FullyQualifiedResourceId workspaceParentFqId =
         new FullyQualifiedResourceId()
             .resourceId(resource.getWorkspaceId().toString())
             .resourceTypeName(SamConstants.SAM_WORKSPACE_RESOURCE);
+
     CreateResourceRequestV2 resourceRequest =
         new CreateResourceRequestV2()
             .resourceId(resource.getResourceId().toString())
             .parent(workspaceParentFqId);
 
-    addWsmResourceOwnerPolicy(resourceRequest);
-    // Only create policies for private resources. Workspace role permissions are handled through
-    // role-based inheritance in Sam instead. This should expand to include policies for
-    // applications in the future.
-    if (resource.getAccessScope() == AccessScopeType.ACCESS_SCOPE_PRIVATE) {
-      // The assigned user is always the current user for private resources.
-      addPrivateResourcePolicies(resourceRequest, privateIamRoles, assignedUserEmail);
-    }
+    var builder =
+        new ControlledResourceSamPolicyBuilder(
+            this,
+            privateIamRole,
+            assignedUserEmail,
+            userRequest,
+            ControlledResourceCategory.get(resource.getAccessScope(), resource.getManagedBy()));
+    builder.addPolicies(resourceRequest);
 
     try {
+      // We use the user request for the create, but could equally well use the WSM SA.
```

Comment:
One risk is if this is delayed the user request token may be expired (I think this is run asynchronously by Stairway).

Reply:
That is a more general problem. We are in the process of switching over to pets for most everything.